proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,435 @@
+  otherwise throw `AccountDataTooSmall`
+  - Copy the chunk into the program account at the offset shifted by the
+  header size
+
+#### Copy
+
+- Instruction accounts:
+  - `[writable]` The program account to copy to.
+  - `[signer]` The authority of the program.
+  - `[]` The account to copy from.
+- Instruction data:
+  - Enum variant `1u32`
+  - `u32` Byte offset at which to write
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes CUs
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the source account is owned by loader v2, v3 or v4,
+  otherwise throw `InvalidArgument`
+  - and look-up the source header size:
+    - loader-v2: 0 bytes
+    - loader-v3 buffer: 37 bytes
+    - loader-v3 programdata: 45 bytes
+    - loader-v4: 48 bytes
+  - Check that the source end offset (sum of source offset and length) does
+  not exceed the maximum (source account length minus the source header size),
+  otherwise throw `AccountDataTooSmall`
+  - Check that the destination end offset (sum of destination offset and
+  length) does not exceed the maximum (program account length minus the loader-v4
+  header size), otherwise throw `AccountDataTooSmall`
+  - Copy the chunk between the program accounts at the offsets, each shifted by
+  the header size of their loader (account owner) respectively
+
+#### SetProgramLength
+
+- Instruction accounts:
+  - `[writable]` The program account to change the size of.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `2u32`
+  - `u32` The new size after the operation.
+- Behavior:
+  - Charge 32 + new_size_in_bytes CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - If this is an initialization (program account length is too short to
+  contain the header):
+    - the owner of the program account is loader-v4,
+    otherwise throw `InvalidAccountOwner`
+    - the program account is writable, otherwise throw `InvalidArgument`
+    - the provided authority (instruction account at index 1) signed,
+    otherwise throw `MissingRequiredSignature`
+  - If this is not an initialization:
+    - Verify the program account
+    - Check that the status stored in the program account is
+    `NeverBeenDeployed` or `Retracted`,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the new length, otherwise throw `InsufficientFunds`
+  - Set the length of the program account to the requested new size plus
```

Comment:
Is it worth specifying special behavior in CPI to limit the new size to `size_at_start_of_top_level_instruction + 10kb`?

Reply:
I don't think this was addressed